# Python Test

In [1]:
# !pip install seaborn

In [5]:
import seaborn as sns
import numpy as np
import pandas as pd

penguins = sns.load_dataset("penguins")

### Question 1
- Inspect the structure of the penguins dataset.
- Find the species with highest average bill length by sex within each island.

In [6]:
penguins.shape

(344, 7)

In [7]:
penguins.head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,Male
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,Female
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,Female
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,Female


In [8]:
penguins.tail()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
339,Gentoo,Biscoe,NaN,NaN,NaN,NaN,NaN
340,Gentoo,Biscoe,46.8,14.3,215.0,4850.0,Female
341,Gentoo,Biscoe,50.4,15.7,222.0,5750.0,Male
342,Gentoo,Biscoe,45.2,14.8,212.0,5200.0,Female
343,Gentoo,Biscoe,49.9,16.1,213.0,5400.0,Male


In [9]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [10]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64

In [11]:
penguins.duplicated().sum()

0

In [12]:
# Remove missing values
penguins = penguins.dropna()

# Shape after cleaning
print(penguins.shape)

(333, 7)


In [13]:
penguins.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
dtype: int64

#### Q1 Observation:
###### The dataset initially contained 344 rows and 7 columns.
###### Some missing values were present in numerical columns and the sex column.
###### No duplicate rows were found in the dataset.
###### Missing values were removed using dropna(), and the cleaned dataset contained 333 rows and 7 columns.

In [14]:
print(penguins['species'].unique())

print(penguins['sex'].unique())

print(penguins['island'].unique())

['Adelie' 'Chinstrap' 'Gentoo']
['Male' 'Female']
['Torgersen' 'Biscoe' 'Dream']


In [15]:
penguins['species'].value_counts()

species
Adelie       146
Gentoo       119
Chinstrap     68
Name: count, dtype: int64

In [16]:
result = (
    penguins.groupby(['island', 'sex', 'species'])['bill_length_mm']
    .mean()
    .reset_index(name='avg_bill_length')
)

result = (
    result.sort_values( ['island', 'sex', 'avg_bill_length'],
        ascending=[True, True, False]
    ).groupby(['island', 'sex']).head(1)
)

print(result)

      island     sex    species  avg_bill_length
1     Biscoe  Female     Gentoo        45.563793
3     Biscoe    Male     Gentoo        49.473770
5      Dream  Female  Chinstrap        46.573529
7      Dream    Male  Chinstrap        51.094118
8  Torgersen  Female     Adelie        37.554167
9  Torgersen    Male     Adelie        40.586957


#### Q1 part b
###### Chinstrap penguins in Dream island showed the highest average bill length for both male and female categories, with male penguins reaching around 51 mm on average.
###### Gentoo penguins in Biscoe island also had relatively high average bill length values, especially among male penguins.
###### Adelie penguins in Torgersen island showed the lowest average bill length compared to the other groups.

### Question 2
- Create a new column 'mass_to_flipper' as body_mass_g / flipper_length_m
- Store result in a new dataframe

In [17]:
penguins[['body_mass_g', 'flipper_length_mm']].head()

,body_mass_g,flipper_length_mm
0,3750.0,181.0
1,3800.0,186.0
2,3250.0,195.0
4,3450.0,193.0
5,3650.0,190.0


In [18]:
penguins['body_mass_g'] / penguins['flipper_length_mm']

0      20.718232
1      20.430108
2      16.666667
4      17.875648
5      19.210526
         ...    
338    23.014019
340    22.558140
341    25.900901
342    24.528302
343    25.352113
Length: 333, dtype: float64

In [19]:
#Created New Column
penguins['mass_to_flipper'] = ( penguins['body_mass_g'] /penguins['flipper_length_mm']
)

In [20]:
penguins.sort_values('mass_to_flipper', ascending=False).head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex,mass_to_flipper
237,Gentoo,Biscoe,49.2,15.2,221.0,6300.0,Male,28.506787
233,Gentoo,Biscoe,48.4,14.6,213.0,5850.0,Male,27.464789
297,Gentoo,Biscoe,51.1,16.3,220.0,6000.0,Male,27.272727
337,Gentoo,Biscoe,48.8,16.2,222.0,6000.0,Male,27.027027
235,Gentoo,Biscoe,49.3,15.7,217.0,5850.0,Male,26.958525


In [21]:
# Average mass_to_flipper ratio by species
penguins.groupby('species')['mass_to_flipper'].mean()

species
Adelie       19.480371
Chinstrap    19.043758
Gentoo       23.414151
Name: mass_to_flipper, dtype: float64

In [22]:
new_df = penguins.copy()


In [23]:
print(new_df[['body_mass_g', 'flipper_length_mm', 'mass_to_flipper']].head())

   body_mass_g  flipper_length_mm  mass_to_flipper
0       3750.0              181.0        20.718232
1       3800.0              186.0        20.430108
2       3250.0              195.0        16.666667
4       3450.0              193.0        17.875648
5       3650.0              190.0        19.210526


#### Q2 insights
###### A new column called mass_to_flipper was created by dividing body mass by flipper length.
###### This new feature helps compare penguin body mass relative to their flipper size.
###### Penguins with higher values may have higher body mass compared to their flipper length.

### Question 3
- Filter only the rows where species is "Adelie".
- Store results in a new dataframe

In [24]:
print(penguins['species'].unique())

['Adelie' 'Chinstrap' 'Gentoo']


In [25]:
print(penguins['species'].value_counts())

species
Adelie       146
Gentoo       119
Chinstrap     68
Name: count, dtype: int64


In [26]:
# Filter only Adelie species
adelie_df = penguins[penguins['species'] == 'Adelie'].copy()

adelie_df.head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex,mass_to_flipper
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,Male,20.718232
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,Female,20.430108
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,Female,16.666667
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,Female,17.875648
5,Adelie,Torgersen,39.3,20.6,190.0,3650.0,Male,19.210526


In [27]:
penguins[penguins['species'].isin(['Adelie', 'Gentoo'])].head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex,mass_to_flipper
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,Male,20.718232
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,Female,20.430108
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,Female,16.666667
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,Female,17.875648
5,Adelie,Torgersen,39.3,20.6,190.0,3650.0,Male,19.210526


In [28]:
penguins['species'].nunique()

3

In [29]:
penguins[(penguins['species'] == 'Adelie') & (penguins['sex']== 'Male')].head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex,mass_to_flipper
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,Male,20.718232
5,Adelie,Torgersen,39.3,20.6,190.0,3650.0,Male,19.210526
7,Adelie,Torgersen,39.2,19.6,195.0,4675.0,Male,23.974359
13,Adelie,Torgersen,38.6,21.2,191.0,3800.0,Male,19.895288
14,Adelie,Torgersen,34.6,21.1,198.0,4400.0,Male,22.222222


In [30]:
penguins[(penguins['island']== 'Dream') | (penguins['island']== 'Torgersen')].sample(5)

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex,mass_to_flipper
45,Adelie,Dream,39.6,18.8,190.0,4600.0,Male,24.210526
19,Adelie,Torgersen,46.0,21.5,194.0,4200.0,Male,21.649485
152,Chinstrap,Dream,46.5,17.9,192.0,3500.0,Female,18.229167
150,Adelie,Dream,36.0,17.1,187.0,3700.0,Female,19.786096
173,Chinstrap,Dream,48.5,17.5,191.0,3400.0,Male,17.801047


In [31]:
penguins[penguins['species'] != 'Adelie'].sample(5)

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex,mass_to_flipper
283,Gentoo,Biscoe,54.3,15.7,231.0,5650.0,Male,24.458874
195,Chinstrap,Dream,45.5,17.0,196.0,3500.0,Female,17.857143
328,Gentoo,Biscoe,43.3,14.0,208.0,4575.0,Female,21.995192
170,Chinstrap,Dream,46.4,18.6,190.0,3450.0,Female,18.157895
340,Gentoo,Biscoe,46.8,14.3,215.0,4850.0,Female,22.558140


#### Q3 Insights
###### The dataset was filtered to include only Adelie penguins and stored in a separate dataframe 'adelie_df'.
###### Adelie penguins are present across multiple islands such as Torgersen, Dream, and Biscoe.
###### The filtered dataframe can now be used for more focused analysis on Adelie penguin.

### Question 4
- Calculate the average bill_length_mm for each species
- Store the results in a new dataframe

In [32]:
print(penguins['species'].unique())

['Adelie' 'Chinstrap' 'Gentoo']


In [33]:
print(penguins[['species', 'bill_length_mm']].head())

  species  bill_length_mm
0  Adelie            39.1
1  Adelie            39.5
2  Adelie            40.3
4  Adelie            36.7
5  Adelie            39.3


In [34]:
# Calculate average bill length for each species
avg_bill_df = (
    penguins.groupby('species')['bill_length_mm']
    .mean().reset_index(name='avg_bill_length_mm')
    
)

print(avg_bill_df)

     species  avg_bill_length_mm
0     Adelie           38.823973
1  Chinstrap           48.833824
2     Gentoo           47.568067


In [35]:
new = penguins.groupby('species')['bill_length_mm'].mean().reset_index(name= "avg_bill").sort_values('avg_bill')
print(new)

     species   avg_bill
0     Adelie  38.823973
2     Gentoo  47.568067
1  Chinstrap  48.833824


In [36]:
penguins.groupby('species')['bill_length_mm'].mean().reset_index(name='avg_bill').sort_values('avg_bill',ascending=False)   

,species,avg_bill
1,Chinstrap,48.833824
2,Gentoo,47.568067
0,Adelie,38.823973


In [37]:
avg_bill_df.sort_values('avg_bill_length_mm', ascending= False)

,species,avg_bill_length_mm
1,Chinstrap,48.833824
2,Gentoo,47.568067
0,Adelie,38.823973


In [38]:
penguins.groupby('species')['bill_length_mm'].agg(['mean','count'])

,mean,count
species,,
Adelie,38.823973,146
Chinstrap,48.833824,68
Gentoo,47.568067,119


##### Q4 Insights
###### Chinstrap penguins showed the highest average bill length among all species, with an average of around 48.8 mm.
###### Gentoo penguins also had relatively high average bill length values compared to Adelie penguins.
###### Adelie penguins had the smallest average bill length among the three species.

### Question 5
- Create a wide-format table showing the number of penguins of each species on each island.
    - Rows: islands
    - Columns: species names
    - Values: count of non-NA observations per species on that island


In [39]:
print(penguins['island'].unique())

['Torgersen' 'Biscoe' 'Dream']


In [40]:
print(penguins['species'].unique())

['Adelie' 'Chinstrap' 'Gentoo']


In [41]:
# Create wide-format table
penguin_count_df = pd.pivot_table(
    penguins,
    index='island',
    columns='species',
    values='bill_length_mm',
    aggfunc='count'
).fillna(0).astype(int)

print(penguin_count_df)

species    Adelie  Chinstrap  Gentoo
island                              
Biscoe         44          0     119
Dream          55         68       0
Torgersen      47          0       0


In [42]:
penguins_count_df = pd.pivot_table(
    penguins,
    index='island',
    columns='species',
    values='bill_length_mm',
    aggfunc='mean'
    ).fillna(0).astype(int)
print(penguins_count_df)

species    Adelie  Chinstrap  Gentoo
island                              
Biscoe         38          0      47
Dream          38         48       0
Torgersen      39          0       0


In [43]:
# validating total penguin count
print(penguin_count_df.sum().sum())

333


#### Q5 Insights
###### Gentoo penguins were found only in Biscoe island, while Chinstrap penguins were present only in Dream island.
###### Adelie penguins were distributed across all three islands: Biscoe, Dream, and Torgersen.
###### The total penguin count from the wide-format table matched the cleaned dataset size of 333, confirming the aggregation was correct.

### Question 6
- Write a function dominant_species_per_island() that:
    - For each island, finds the species with the highest number of observations (non-NA species only).
    - If there's a tie, return all tied species.
    - Return a dataframe with columns: island, species, count

In [44]:
print(penguins['species'].value_counts())

species
Adelie       146
Gentoo       119
Chinstrap     68
Name: count, dtype: int64


In [45]:
print(
    penguins.groupby(['island', 'species'])
    .size()
)

island     species  
Biscoe     Adelie        44
           Gentoo       119
Dream      Adelie        55
           Chinstrap     68
Torgersen  Adelie        47
dtype: int64


In [46]:
def dominant_species_per_island(df):
    species_count = df.groupby(['island', 'species']).size().reset_index(name='count')

    max_count = species_count.groupby('island')['count'].transform('max')

    result = species_count[species_count['count'] == max_count]

    return result

dominant_df = dominant_species_per_island(penguins)

print(dominant_df)

      island    species  count
1     Biscoe     Gentoo    119
3      Dream  Chinstrap     68
4  Torgersen     Adelie     47


#### Q6 Insights 
###### The analysis shows that different penguin species are more commonly found in specific islands.
###### Gentoo penguins were observed the most in Biscoe island, while Chinstrap penguins were more common in Dream island.
###### Torgersen island mainly contained Adelie penguins, showing a clear difference in species distribution across islands.

### Question 7
We've prepared a synthetic dataset for you simulating an index driven by random growth rates.
- Your task is to reconstruct the index using only the starting value and the given growth rates.

In [47]:
np.random.seed(42)
index_data = pd.DataFrame({
    "date": pd.date_range(start="2000-01-01", periods=24, freq="MS"),
    "growth_rate": np.random.uniform(-0.02, 0.02, 24),
    "index": np.nan
})
index_data.iloc[0, index_data.columns.get_loc("index")] = 100

In [48]:
index_data.head()

,date,growth_rate,index
0,2000-01-01,-0.005018,100.0
1,2000-02-01,0.018029,NaN
2,2000-03-01,0.009280,NaN
3,2000-04-01,0.003946,NaN
4,2000-05-01,-0.013759,NaN


In [49]:
index_data.iloc[0]

date           2000-01-01 00:00:00
growth_rate              -0.005018
index                        100.0
Name: 0, dtype: object

In [50]:
for i in range(1, len(index_data)):
    previous_value = index_data.loc[i - 1, 'index']
    growth = index_data.loc[i, 'growth_rate']

    index_data.loc[i, 'index'] = previous_value * (1 + growth)

index_data.head()

,date,growth_rate,index
0,2000-01-01,-0.005018,100.000000
1,2000-02-01,0.018029,101.802857
2,2000-03-01,0.009280,102.747563
3,2000-04-01,0.003946,103.153040
4,2000-05-01,-0.013759,101.733731


#### Q7 Insights
###### The index started with a base value of 100 and changed month by month using the given growth rates.
###### Positive growth rates increased the index value, while negative growth rates caused small decreases.
###### Each new index value was calculated using the previous month’s index value.
###### This helped recreate the complete index series from the starting value and growth data.

In [51]:
x = 10 
name = "Anuj"
price =99.5
is_valid = True

In [52]:
a=[1,2,3,4,5]
a.append(6)
a.append([7,8])
print(a)

[1, 2, 3, 4, 5, 6, [7, 8]]


In [53]:
a = [1,2]
b = [1,2]
print(a == b)
print(a is b)

True
False


In [54]:
s = "anuj"
print(s[::-1])

juna


In [55]:
s = "Banana"
print(s.count('a'))

3


In [56]:
s= "madam"
print(s == s[::-1])

True


In [57]:
for i in range (1,11):
    if i % 2 == 0:
        print(i)

2
4
6
8
10


In [58]:
nums = [1, 2, 3, 4, 5]
print(sum(nums))
        

15


In [59]:
text = "banana"
freq = {}
for ch in text :
    freq[ch] = freq.get(ch, 0)+1
print(freq)

{'b': 1, 'a': 3, 'n': 2}


In [60]:
def add(a,b):
    return a+b
print(add(3,5))

8


In [61]:
square = lambda x: x**2
print(square(3))

9


In [77]:
emt = pd.read_excel(r"C:\Users\anuj1\OneDrive\Documents\Test_Data_Analyst.xlsx",sheet_name=None)


In [ ]:
print(emt.keys())

dict_keys(['MarketData', 'Population', 'FX_Rates', 'Q1', 'Q2', 'Q3'])


In [78]:
market_df = emt['MarketData']
population_df = emt['Population']
fx_df = emt['FX_Rates']
q1_df = emt['Q1']
q2_df = emt['Q2']
q3_df = emt['Q3']

In [79]:
market_df.head()

,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,...,Unnamed: 11,Unnamed: 12,Unnamed: 13,Unnamed: 14,Unnamed: 15,Unnamed: 16,Unnamed: 17,Unnamed: 18,Unnamed: 19,Unnamed: 20
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,Region,Country,Industry,Category,Sub-Category,Data Type,Unit,2018.0,2019.0,...,2021.0,2022.0,2023.0,2024.0,2025.0,2026.0,2027.0,2028.0,Historic CAGR % (2018-2023),Forecast CAGR % (2023-2028)
2,NaN,Asia Pacific,India,Beauty and Personal Care,Skin Care,Facial Care,Retail Value RSP,USD million,759.0,824.4,...,896.2,994.2,909.9,916.3,922.9,944.7,995.4,1134.0,3.69,4.5
3,NaN,Asia Pacific,India,Beauty and Personal Care,Skin Care,Facial Care,Retail Volume,Tonnes,96.7,91.0,...,91.6,86.5,103.6,112.7,120.3,120.3,117.1,115.0,1.39,2.11
4,NaN,Asia Pacific,India,Beauty and Personal Care,Skin Care,Body Care,Retail Value RSP,USD million,346.5,326.5,...,369.7,397.9,326.6,329.7,346.6,327.1,313.3,313.6,-1.18,-0.81


In [80]:
market_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 898 entries, 0 to 897
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   0 non-null      float64
 1   Unnamed: 1   897 non-null    object 
 2   Unnamed: 2   897 non-null    object 
 3   Unnamed: 3   897 non-null    object 
 4   Unnamed: 4   897 non-null    object 
 5   Unnamed: 5   897 non-null    object 
 6   Unnamed: 6   897 non-null    object 
 7   Unnamed: 7   897 non-null    object 
 8   Unnamed: 8   897 non-null    float64
 9   Unnamed: 9   897 non-null    float64
 10  Unnamed: 10  897 non-null    float64
 11  Unnamed: 11  897 non-null    float64
 12  Unnamed: 12  897 non-null    float64
 13  Unnamed: 13  897 non-null    float64
 14  Unnamed: 14  897 non-null    float64
 15  Unnamed: 15  897 non-null    float64
 16  Unnamed: 16  897 non-null    float64
 17  Unnamed: 17  897 non-null    float64
 18  Unnamed: 18  897 non-null    float64
 19  Unnamed:

In [81]:
market_df.columns

Index(['Unnamed: 0', 'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4',
       'Unnamed: 5', 'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9',
       'Unnamed: 10', 'Unnamed: 11', 'Unnamed: 12', 'Unnamed: 13',
       'Unnamed: 14', 'Unnamed: 15', 'Unnamed: 16', 'Unnamed: 17',
       'Unnamed: 18', 'Unnamed: 19', 'Unnamed: 20'],
      dtype='object')

In [82]:
market_df= market_df.dropna(how='all')

In [83]:
market_df = market_df.iloc[:,1:]


In [84]:
market_df = market_df.reset_index(drop=True)

In [85]:
market_df.columns = market_df.iloc[0]

In [86]:
market_df = market_df[1:]

In [87]:
market_df = market_df.reset_index(drop=True)

In [88]:
market_df.head()

,Region,Country,Industry,Category,Sub-Category,Data Type,Unit,2018.0,2019.0,2020.0,2021.0,2022.0,2023.0,2024.0,2025.0,2026.0,2027.0,2028.0,Historic CAGR % (2018-2023),Forecast CAGR % (2023-2028)
0,Asia Pacific,India,Beauty and Personal Care,Skin Care,Facial Care,Retail Value RSP,USD million,759.0,824.4,789.6,896.2,994.2,909.9,916.3,922.9,944.7,995.4,1134.0,3.69,4.5
1,Asia Pacific,India,Beauty and Personal Care,Skin Care,Facial Care,Retail Volume,Tonnes,96.7,91.0,95.0,91.6,86.5,103.6,112.7,120.3,120.3,117.1,115.0,1.39,2.11
2,Asia Pacific,India,Beauty and Personal Care,Skin Care,Body Care,Retail Value RSP,USD million,346.5,326.5,353.1,369.7,397.9,326.6,329.7,346.6,327.1,313.3,313.6,-1.18,-0.81
3,Asia Pacific,India,Beauty and Personal Care,Skin Care,Body Care,Retail Volume,Tonnes,83.8,99.5,114.0,117.9,117.5,124.2,134.3,145.7,158.8,158.0,157.0,8.19,4.8
4,Asia Pacific,India,Beauty and Personal Care,Hair Care,Shampoo,Retail Value RSP,USD million,1059.3,1074.9,1021.9,1081.7,1154.4,1111.7,1203.4,1257.7,1319.2,1366.8,1357.0,0.97,4.07


In [89]:
market_df.columns

Index([                     'Region',                     'Country',
                          'Industry',                    'Category',
                      'Sub-Category',                   'Data Type',
                              'Unit',                        2018.0,
                              2019.0,                        2020.0,
                              2021.0,                        2022.0,
                              2023.0,                        2024.0,
                              2025.0,                        2026.0,
                              2027.0,                        2028.0,
       'Historic CAGR % (2018-2023)', 'Forecast CAGR % (2023-2028)'],
      dtype='object', name=0)

In [90]:
population_df.head()

,Unnamed: 0,Unnamed: 1,Unnamed: 2
0,NaN,NaN,NaN
1,NaN,Country,"Population, 2023 (million)"
2,NaN,India,1428.6
3,NaN,China,1411.8
4,NaN,Japan,124.7


In [91]:
population_df = population_df.iloc[1:,1:]

population_df = population_df.reset_index(drop=True)

population_df.head()



,Unnamed: 1,Unnamed: 2
0,Country,"Population, 2023 (million)"
1,India,1428.6
2,China,1411.8
3,Japan,124.7
4,Indonesia,277.5


In [92]:
population_df.columns = population_df.iloc[0]
population_df = population_df[1:]
population_df = population_df.reset_index(drop=True)


In [93]:
population_df.head()

,Country,"Population, 2023 (million)"
0,India,1428.6
1,China,1411.8
2,Japan,124.7
3,Indonesia,277.5
4,USA,334.9


In [94]:
fx_df.head()

,Unnamed: 0,Unnamed: 1,Unnamed: 2,Unnamed: 3
0,NaN,NaN,NaN,NaN
1,NaN,Country,Local Currency,USD per 1 LCU (2023)
2,NaN,India,INR,0.0121
3,NaN,China,CNY,0.1411
4,NaN,Japan,JPY,0.0071


In [95]:
fx_df = fx_df.iloc[1:,1:]
fx_df = fx_df.reset_index(drop=True)
fx_df.head()

,Unnamed: 1,Unnamed: 2,Unnamed: 3
0,Country,Local Currency,USD per 1 LCU (2023)
1,India,INR,0.0121
2,China,CNY,0.1411
3,Japan,JPY,0.0071
4,Indonesia,IDR,0.000066


In [96]:
fx_df.columns= fx_df.iloc[0]
fx_df = fx_df[1:]
fx_df =fx_df.reset_index(drop=True)
fx_df.head()

,Country,Local Currency,USD per 1 LCU (2023)
0,India,INR,0.0121
1,China,CNY,0.1411
2,Japan,JPY,0.0071
3,Indonesia,IDR,0.000066
4,USA,USD,1


In [97]:
retail_value_df = market_df[
    market_df['Data Type'] == 'Retail Value RSP'
]

In [98]:
retail_value_df.head()

,Region,Country,Industry,Category,Sub-Category,Data Type,Unit,2018.0,2019.0,2020.0,2021.0,2022.0,2023.0,2024.0,2025.0,2026.0,2027.0,2028.0,Historic CAGR % (2018-2023),Forecast CAGR % (2023-2028)
0,Asia Pacific,India,Beauty and Personal Care,Skin Care,Facial Care,Retail Value RSP,USD million,759.0,824.4,789.6,896.2,994.2,909.9,916.3,922.9,944.7,995.4,1134.0,3.69,4.5
2,Asia Pacific,India,Beauty and Personal Care,Skin Care,Body Care,Retail Value RSP,USD million,346.5,326.5,353.1,369.7,397.9,326.6,329.7,346.6,327.1,313.3,313.6,-1.18,-0.81
4,Asia Pacific,India,Beauty and Personal Care,Hair Care,Shampoo,Retail Value RSP,USD million,1059.3,1074.9,1021.9,1081.7,1154.4,1111.7,1203.4,1257.7,1319.2,1366.8,1357.0,0.97,4.07
6,Asia Pacific,India,Beauty and Personal Care,Hair Care,Conditioner,Retail Value RSP,USD million,1208.7,1232.2,1235.2,1194.5,1133.1,1254.0,1289.7,1225.9,1195.5,1193.5,1065.5,0.74,-3.21
8,Asia Pacific,India,Beauty and Personal Care,Colour Cosmetics,Lip Products,Retail Value RSP,USD million,610.7,666.1,737.6,723.5,778.3,827.9,869.7,859.0,907.6,922.2,941.3,6.27,2.6


In [99]:
country_value = retail_value_df.groupby('Country')[2023.0].sum().reset_index(name='total_retail_value_2023')
country_value.head(25)

,Country,total_retail_value_2023
0,Argentina,13395.6
1,Brazil,39874.6
2,Canada,38771.5
3,Chile,10954.6
4,China,108822.2
5,Colombia,10070.1
6,France,54771.1
7,Germany,53692.4
8,India,32165.7
9,Indonesia,24533.9


In [100]:
print(country_value)

           Country  total_retail_value_2023
0        Argentina                  13395.6
1           Brazil                  39874.6
2           Canada                  38771.5
3            Chile                  10954.6
4            China                 108822.2
5         Colombia                  10070.1
6           France                  54771.1
7          Germany                  53692.4
8            India                  32165.7
9        Indonesia                  24533.9
10           Italy                  46468.5
11           Japan                  74703.9
12          Mexico                  26559.5
13           Spain                  27351.5
14             USA                 130565.2
15  United Kingdom                  37142.5


In [101]:
country_value.tail()

,Country,total_retail_value_2023
11,Japan,74703.9
12,Mexico,26559.5
13,Spain,27351.5
14,USA,130565.2
15,United Kingdom,37142.5


In [102]:
country_value.shape

(16, 2)

In [103]:

country_value.sample(16)

,Country,total_retail_value_2023
9,Indonesia,24533.9
1,Brazil,39874.6
5,Colombia,10070.1
0,Argentina,13395.6
10,Italy,46468.5
7,Germany,53692.4
4,China,108822.2
12,Mexico,26559.5
8,India,32165.7
3,Chile,10954.6


In [104]:
merged_df = pd.merge(country_value,population_df,on='Country', how='left')

In [105]:
print(merged_df)

           Country  total_retail_value_2023 Population, 2023 (million)
0        Argentina                  13395.6                       45.7
1           Brazil                  39874.6                      210.7
2           Canada                  38771.5                       39.6
3            Chile                  10954.6                       19.5
4            China                 108822.2                     1411.8
5         Colombia                  10070.1                       51.9
6           France                  54771.1                       65.8
7          Germany                  53692.4                       84.4
8            India                  32165.7                     1428.6
9        Indonesia                  24533.9                      277.5
10           Italy                  46468.5                       58.9
11           Japan                  74703.9                      124.7
12          Mexico                  26559.5                      128.4
13    

In [106]:
merged_df['per_capita']= merged_df['total_retail_value_2023']/ merged_df['Population, 2023 (million)']

In [107]:
merged_df.head()

,Country,total_retail_value_2023,"Population, 2023 (million)",per_capita
0,Argentina,13395.6,45.7,293.12035
1,Brazil,39874.6,210.7,189.24822
2,Canada,38771.5,39.6,979.078283
3,Chile,10954.6,19.5,561.774359
4,China,108822.2,1411.8,77.080465


In [108]:
region_df = market_df[['Country','Region']].drop_duplicates()

In [109]:
print(region_df)

0           Country          Region
0             India    Asia Pacific
56            China    Asia Pacific
112           Japan    Asia Pacific
168       Indonesia    Asia Pacific
224             USA   North America
280          Canada   North America
336  United Kingdom  Western Europe
392         Germany  Western Europe
448          France  Western Europe
504           Italy  Western Europe
560           Spain  Western Europe
616          Brazil   Latin America
672          Mexico   Latin America
728       Argentina   Latin America
784           Chile   Latin America
840        Colombia   Latin America


In [110]:
beauty_df =market_df[market_df['Industry'] == 'Beauty and Personal Care']

In [111]:
beauty_value_df = beauty_df[beauty_df['Data Type']== 'Retail Value RSP']

In [112]:
beauty_county_df = beauty_value_df.groupby('Country')[2023.0].sum().reset_index(name='total_beauty_retail_value_2023')
print(beauty_county_df)

           Country  total_beauty_retail_value_2023
0        Argentina                          2982.4
1           Brazil                         11817.8
2           Canada                          7444.8
3            Chile                          2257.2
4            China                         27578.0
5         Colombia                          1998.8
6           France                         10147.4
7          Germany                         10393.3
8            India                          8237.8
9        Indonesia                          5737.1
10           Italy                          7137.9
11           Japan                         13192.1
12          Mexico                          4233.7
13           Spain                          5125.6
14             USA                         26862.8
15  United Kingdom                          8286.9


In [113]:
beauty_volume_df = beauty_df[beauty_df['Data Type'] == 'Retail Volume']

In [114]:
beauty_country_volume = beauty_volume_df.groupby('Country')[2023.0].sum().reset_index(name='total_beauty_retail_volume_2023')
print(beauty_country_volume)

           Country  total_beauty_retail_volume_2023
0        Argentina                            544.1
1           Brazil                           1333.9
2           Canada                           1845.5
3            Chile                            278.5
4            China                           3977.4
5         Colombia                            322.7
6           France                           1740.6
7          Germany                           1953.6
8            India                            995.5
9        Indonesia                            810.7
10           Italy                           1664.7
11           Japan                           2549.9
12          Mexico                            797.3
13           Spain                            761.9
14             USA                           4913.1
15  United Kingdom                           1940.4


In [115]:
merged_df = pd.merge(beauty_county_df, population_df, on='Country',how='left')
print(merged_df)

           Country  total_beauty_retail_value_2023 Population, 2023 (million)
0        Argentina                          2982.4                       45.7
1           Brazil                         11817.8                      210.7
2           Canada                          7444.8                       39.6
3            Chile                          2257.2                       19.5
4            China                         27578.0                     1411.8
5         Colombia                          1998.8                       51.9
6           France                         10147.4                       65.8
7          Germany                         10393.3                       84.4
8            India                          8237.8                     1428.6
9        Indonesia                          5737.1                      277.5
10           Italy                          7137.9                       58.9
11           Japan                         13192.1              

In [116]:
merged_df['per_capita_beauty'] = merged_df['total_beauty_retail_value_2023'] / merged_df['Population, 2023 (million)']
print(merged_df[['Country', 'per_capita_beauty']])

           Country per_capita_beauty
0        Argentina         65.260394
1           Brazil         56.088277
2           Canada             188.0
3            Chile        115.753846
4            China         19.533928
5         Colombia         38.512524
6           France        154.215805
7          Germany        123.143365
8            India          5.766345
9        Indonesia         20.674234
10           Italy        121.186757
11           Japan        105.790698
12          Mexico         32.972741
13           Spain        106.561331
14             USA         80.211406
15  United Kingdom        122.406204


In [117]:
merged_df['rank'] = merged_df['per_capita_beauty'].rank(ascending=False)

In [118]:
merged_df = merged_df.sort_values('rank')
merged_df.head(10)

,Country,total_beauty_retail_value_2023,"Population, 2023 (million)",per_capita_beauty,rank
2,Canada,7444.8,39.6,188.0,1.0
6,France,10147.4,65.8,154.215805,2.0
7,Germany,10393.3,84.4,123.143365,3.0
15,United Kingdom,8286.9,67.7,122.406204,4.0
10,Italy,7137.9,58.9,121.186757,5.0
3,Chile,2257.2,19.5,115.753846,6.0
13,Spain,5125.6,48.1,106.561331,7.0
11,Japan,13192.1,124.7,105.790698,8.0
14,USA,26862.8,334.9,80.211406,9.0
0,Argentina,2982.4,45.7,65.260394,10.0


In [119]:
merged_df['Country']

2             Canada
6             France
7            Germany
15    United Kingdom
10             Italy
3              Chile
13             Spain
11             Japan
14               USA
0          Argentina
1             Brazil
5           Colombia
12            Mexico
9          Indonesia
4              China
8              India
Name: Country, dtype: object

In [120]:
sleceted_countries = ['USA','China','Germany','United Kingdom','Japan','France','India','Brazil','Italy','Spain']
filtered_df = merged_df[merged_df['Country'].isin(sleceted_countries)]
filtered_df                     


,Country,total_beauty_retail_value_2023,"Population, 2023 (million)",per_capita_beauty,rank
6,France,10147.4,65.8,154.215805,2.0
7,Germany,10393.3,84.4,123.143365,3.0
15,United Kingdom,8286.9,67.7,122.406204,4.0
10,Italy,7137.9,58.9,121.186757,5.0
13,Spain,5125.6,48.1,106.561331,7.0
11,Japan,13192.1,124.7,105.790698,8.0
14,USA,26862.8,334.9,80.211406,9.0
1,Brazil,11817.8,210.7,56.088277,11.0
4,China,27578.0,1411.8,19.533928,15.0
8,India,8237.8,1428.6,5.766345,16.0


In [121]:
filtered_df.columns

Index(['Country', 'total_beauty_retail_value_2023',
       'Population, 2023 (million)', 'per_capita_beauty', 'rank'],
      dtype='object')

In [122]:
filtered_df['rank'] = filtered_df['per_capita_beauty'].rank(ascending=False)
filtered_df

C:\Users\anuj1\AppData\Local\Temp\ipykernel_22736\2294691640.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df['rank'] = filtered_df['per_capita_beauty'].rank(ascending=False)


,Country,total_beauty_retail_value_2023,"Population, 2023 (million)",per_capita_beauty,rank
6,France,10147.4,65.8,154.215805,1.0
7,Germany,10393.3,84.4,123.143365,2.0
15,United Kingdom,8286.9,67.7,122.406204,3.0
10,Italy,7137.9,58.9,121.186757,4.0
13,Spain,5125.6,48.1,106.561331,5.0
11,Japan,13192.1,124.7,105.790698,6.0
14,USA,26862.8,334.9,80.211406,7.0
1,Brazil,11817.8,210.7,56.088277,8.0
4,China,27578.0,1411.8,19.533928,9.0
8,India,8237.8,1428.6,5.766345,10.0


In [123]:
fx_df

,Country,Local Currency,USD per 1 LCU (2023)
0,India,INR,0.0121
1,China,CNY,0.1411
2,Japan,JPY,0.0071
3,Indonesia,IDR,0.000066
4,USA,USD,1
5,Canada,CAD,0.7409
6,United Kingdom,GBP,1.2434
7,Germany,EUR,1.0813
8,France,EUR,1.0813
9,Italy,EUR,1.0813


In [124]:
fx_df.columns

Index(['Country', 'Local Currency', 'USD per 1 LCU (2023)'], dtype='object', name=0)

In [125]:
fx_merged_df = pd.merge(filtered_df, fx_df, on='Country', how='left')
fx_merged_df

,Country,total_beauty_retail_value_2023,"Population, 2023 (million)",per_capita_beauty,rank,Local Currency,USD per 1 LCU (2023)
0,France,10147.4,65.8,154.215805,1.0,EUR,1.0813
1,Germany,10393.3,84.4,123.143365,2.0,EUR,1.0813
2,United Kingdom,8286.9,67.7,122.406204,3.0,GBP,1.2434
3,Italy,7137.9,58.9,121.186757,4.0,EUR,1.0813
4,Spain,5125.6,48.1,106.561331,5.0,EUR,1.0813
5,Japan,13192.1,124.7,105.790698,6.0,JPY,0.0071
6,USA,26862.8,334.9,80.211406,7.0,USD,1
7,Brazil,11817.8,210.7,56.088277,8.0,BRL,0.2002
8,China,27578.0,1411.8,19.533928,9.0,CNY,0.1411
9,India,8237.8,1428.6,5.766345,10.0,INR,0.0121


In [126]:
fx_merged_df['local_currency_per_mn'] = fx_merged_df['total_beauty_retail_value_2023'] / fx_merged_df['USD per 1 LCU (2023)']
fx_merged_df

,Country,total_beauty_retail_value_2023,"Population, 2023 (million)",per_capita_beauty,rank,Local Currency,USD per 1 LCU (2023),local_currency_per_mn
0,France,10147.4,65.8,154.215805,1.0,EUR,1.0813,9384.44465
1,Germany,10393.3,84.4,123.143365,2.0,EUR,1.0813,9611.856099
2,United Kingdom,8286.9,67.7,122.406204,3.0,GBP,1.2434,6664.709667
3,Italy,7137.9,58.9,121.186757,4.0,EUR,1.0813,6601.220753
4,Spain,5125.6,48.1,106.561331,5.0,EUR,1.0813,4740.220105
5,Japan,13192.1,124.7,105.790698,6.0,JPY,0.0071,1858042.253521
6,USA,26862.8,334.9,80.211406,7.0,USD,1,26862.8
7,Brazil,11817.8,210.7,56.088277,8.0,BRL,0.2002,59029.97003
8,China,27578.0,1411.8,19.533928,9.0,CNY,0.1411,195450.035436
9,India,8237.8,1428.6,5.766345,10.0,INR,0.0121,680809.917355


In [172]:
market_df.to_csv(r"C:\Users\anuj1\OneDrive\Documents\market_clean.csv",index=False)

population_df.to_csv(r"C:\Users\anuj1\OneDrive\Documents\population_clean.csv",index=False)

fx_df.to_csv(r"C:\Users\anuj1\OneDrive\Documents\fx_clean.csv",index=False)

NameError: name 'market_df' is not defined

In [127]:
market_df.head()

,Region,Country,Industry,Category,Sub-Category,Data Type,Unit,2018.0,2019.0,2020.0,2021.0,2022.0,2023.0,2024.0,2025.0,2026.0,2027.0,2028.0,Historic CAGR % (2018-2023),Forecast CAGR % (2023-2028)
0,Asia Pacific,India,Beauty and Personal Care,Skin Care,Facial Care,Retail Value RSP,USD million,759.0,824.4,789.6,896.2,994.2,909.9,916.3,922.9,944.7,995.4,1134.0,3.69,4.5
1,Asia Pacific,India,Beauty and Personal Care,Skin Care,Facial Care,Retail Volume,Tonnes,96.7,91.0,95.0,91.6,86.5,103.6,112.7,120.3,120.3,117.1,115.0,1.39,2.11
2,Asia Pacific,India,Beauty and Personal Care,Skin Care,Body Care,Retail Value RSP,USD million,346.5,326.5,353.1,369.7,397.9,326.6,329.7,346.6,327.1,313.3,313.6,-1.18,-0.81
3,Asia Pacific,India,Beauty and Personal Care,Skin Care,Body Care,Retail Volume,Tonnes,83.8,99.5,114.0,117.9,117.5,124.2,134.3,145.7,158.8,158.0,157.0,8.19,4.8
4,Asia Pacific,India,Beauty and Personal Care,Hair Care,Shampoo,Retail Value RSP,USD million,1059.3,1074.9,1021.9,1081.7,1154.4,1111.7,1203.4,1257.7,1319.2,1366.8,1357.0,0.97,4.07


In [128]:
population_df

,Country,"Population, 2023 (million)"
0,India,1428.6
1,China,1411.8
2,Japan,124.7
3,Indonesia,277.5
4,USA,334.9
5,Canada,39.6
6,United Kingdom,67.7
7,Germany,84.4
8,France,65.8
9,Italy,58.9


In [129]:
fx_df

,Country,Local Currency,USD per 1 LCU (2023)
0,India,INR,0.0121
1,China,CNY,0.1411
2,Japan,JPY,0.0071
3,Indonesia,IDR,0.000066
4,USA,USD,1
5,Canada,CAD,0.7409
6,United Kingdom,GBP,1.2434
7,Germany,EUR,1.0813
8,France,EUR,1.0813
9,Italy,EUR,1.0813
